Release Notes 1.15.0 (#878)
crystal-lang/crystal-website: _releases/2025-01-09-1.15.0-released.md

improving the language! ❤️

## Changes

Below we list the most remarkable changes in the language, compiler and stdlib.
For more details, visit the [full changelog](https://github.com/crystal-lang/crystal/releases/tag/1.15.0).

### Breaking

Constants are allowed to start with non-ascii uppercase and titlecase letters.
Previously, only ASCII uppercase letters were valid as first character for the
name of a constant.

This change makes things right, but it could break code that was valid before:
Identifiers starting with non-ascii uppercase or titlecase letters were
previously considered variable names, but are now constant names.
The following example compiles in 1.14 but errors in 1.15:

```cr
Á = 1
Á = 2 # Error: already initialized constant Á
```

Read more in [#15148].

_Thanks , [@nanobowers]_

[#15148]: https://github.com/crystal-lang/crystal/#15148

### Standard library

[`Iterator(T).empty`] creates an iterator with no elements ([#15039])

_Thanks [@spuun]_

A new method [`Enumerable#find_value`] returns the first truthy block result ([#14893]).

```cr
["", nil, "foo", "bar"].find_value(&.presence) # => "foo"
```

_Thanks [@jgaskins]_

> `Log` messages do not emit when issues with a block and the block outputs `nil`.
> This was incorrectly also the case even when an `exception` instance was passed
> to the emit method. This was now fixed in [#15253] and message always emit when
> an exception is passed. And we added new overloads for passing an exception
> without giving a block ([#15257]).

Reviewer: ```suggestion
`Log` messages do not emit when issued with a block and the block returns `nil`.
This was incorrectly also the case even when an `exception` instance was passed
to the emit method. This was fixed in [#15253] and a message is now emitted when
an exception is passed. We also added new overloads for passing an exception
without giving a block ([#15257]).
```

Refactor Lifetime Event Loop ([#14996], [#15205], [#15206], [#15215])

In the process we also refactored the `Crystal::EventLoop` namespace ([#15226]),
and some details of the `libevent` ([#15243]) and IOCP event loop drivers ([#15238]).

This effort is part of the ongoing [project to improve multi-threading support](/2024/02/09/84codes-manas-mt)
with the help of [84codes].
We've started to integrate multi-threaded execution contexts from [`ysbaddaden/execution_context`](https://github.com/ysbaddaden/execution_context)
into the standard library, implementing [RFC 2](https://github.com/crystal-lang/rfcs/pull/2).
We expect this to become available in the next release.

_Thanks [@ysbaddaden]_

#### Stabilized platform support

In the process of testing the event loop implementations across platforms we
updated specs and fixed some smaller issues with platform support on operating
systems without automated continuous testing (OpenBSD, FreeBSD, Dragonfly BSD
and NetBSD).

We also enabled indirect branch tracking on OpenBSD where this is required
([#15122]). On other systems it's opt in via compiler flags:

- `-Dcf-protection=branch` for x86_64
- `-Dcf-protection=return` for x86_64

> - `-Dcf-protection==full` for x86_64

Reviewer: ```suggestion
- `-Dcf-protection=full` for x86_64
```

and our official distribution channels.
See [crystal-lang.org/install](https://crystal-lang.org/install/) for
installation instructions.

## Stats

This release includes [161 changes since 1.14.1](https://github.com/crystal-lang/crystal/pulls?q=is%3Apr+milestone%3A1.15.0)
by 25 contributors.  We thank all the contributors for all the effort put into
improving the language! ❤️

## Changes

Below we list the most remarkable changes in the language, compiler and stdlib.
For more details, visit the [full changelog](https://github.com/crystal-lang/crystal/releases/tag/1.15.0).

### Breaking

Constants are allowed to start with non-ascii uppercase and titlecase letters.
Previously, only ASCII uppercase letters were valid as first character for the
name of a constant.

This change makes things right, but it could break code that was valid before:
Identifiers starting with non-ascii uppercase or titlecase letters were
previously considered variable names, but are now constant names.
The following example compiles in 1.14 but errors in 1.15:

```cr
Á = 1
Á = 2 # Error: already initialized constant Á
```

Read more in [#15148].

_Thanks , [@nanobowers]_

> [#15148]: https://github.com/crystal-lang/crystal/#15148

Reviewer 2: ```suggestion
[#15148]: https://github.com/crystal-lang/crystal/pull/15148
```

#### Lifetime EventLoop

As introduced in the blog post [_A new Event Loop for UNIX operating systems_](/2024/11/05/lifetime-event-loop/)
we have a new event loop driver implementing [RFC #0009].
It integrates directly with the system selectors on Unix systems, removing the
dependency on [`libevent`].

The new implementation is supported on Linux, macOS, FreeBSD and Android and
automatically enabled on these systems and do not require any changes to user
code.
In case you notice any issues with the new implementation, it's possible to fall
back to the old driver with the compile time flag `-Devloop=libevent`.
Read more about [caveats in the blog post][caveats] and [availability in the RFC][availability].

This effort is part of the ongoing [project to improve multi-threading support](/2024/02/09/84codes-manas-mt)
with the help of [84codes].
We've started to integrate multi-threaded execution contexts from [`ysbaddaden/execution_context`]
into the standard library, implementing [RFC 2](https://github.com/crystal-lang/rfcs/pull/2).
We expect this to become available in the next release.

_Thanks [@ysbaddaden]_

> [`libevent`]: (https://libevent.org/)

Reviewer 2: ```suggestion
[`libevent`]: https://libevent.org/
```

> [caveats]: /2024/02/09/84codes-manas-mt#caveats

Reviewer 2: ```suggestion
[caveats]: /2024/11/05/lifetime-event-loop/#caveats
```

[`ysbaddaden/execution_context`]: https://github.com/ysbaddaden/execution_context

#### Stabilized platform support

In the process of testing the event loop implementations across platforms we
updated specs and fixed some smaller issues with platform support on operating
systems without automated continuous testing (OpenBSD, FreeBSD, Dragonfly BSD
and NetBSD).

We also enabled indirect branch tracking on OpenBSD where this is required
([#15122]). On other systems it's opt in via compiler flags:

- `-Dcf-protection=branch` for x86_64
- `-Dcf-protection=return` for x86_64
- `-Dcf-protection=full` for x86_64
- `-Dbranch-protection=bti` for aarch64

_Thanks  [@ysbaddaden]_

[#15122]: https://github.com/crystal-lang/crystal/issues/15122

### Windows

Our Windows efforts made a big step forward with adding MinGW-W64 and MSYS2 as
alternative compiler toolchains.  This required a bit of work for the compiler
tooling and to cleanly separate toolchain and OS portability.  These targets
enable cross-compilation for Windows binaries from Unix systems.

And with the implementation of context switch ([#15155]) we're also making good
progress for Windows on ARM.

> This adds three new [Tier 3 targets](https://crystal-lang.org/reference/1.15.0/syntax_and_semantics/platform_support.html)

Reviewer 2: ```suggestion
This adds three new [Tier 3 targets](https://crystal-lang.org/reference/1.15/syntax_and_semantics/platform_support.html)
```

_Thanks  [@RX14]_

[#15174]: https://github.com/crystal-lang/crystal/issues/15174

### Compiler tools

We enabled some pending formatter features which had been available as opt-in
for a while ([#14718]). These are likely to cause changes in a lot of code bases,
so we collected a bunch of them to activate in a single release.
These changes are backwards-compatible with the formatter from 1.14.1.

The `crystal unreachable` tool has a new `codecov` formatter ([#15059]).

_Thanks  [@Blacksmoke16]_

[#14718]: https://github.com/crystal-lang/crystal/issues/14718
[#15059]: https://github.com/crystal-lang/crystal/issues/15059

### Dependency Updates

We updated the bindings for LibreSSL, enabling a bunch of new functions ([#15177]).

_Thanks  [@straight-shoota]_

[#15177]: https://github.com/crystal-lang/crystal/issues/15177

### Shards 0.19.0

The bundled shards release was updated to [0.19.0](https://github.com/crystal-lang/shards/releases/tag/v0.19.0).

This new shards release brings some small improvements, most notably
it forwards ARGV unmodified to subcommands ([#631]), and supports Codeberg as a
git resolver ([#656]).

_Thanks [@luislavena] and [@miry]_

> [#631]: https://github.com/crystal-lang/crystal/issues/631

Reviewer 2: ```suggestion
[#631]: https://github.com/crystal-lang/shards/pull/631
```

> [#656]: https://github.com/crystal-lang/crystal/issues/656

Reviewer 2: ```suggestion
[#656]: https://github.com/crystal-lang/shards/pull/656
```
